$\qquad$ $\qquad$$\qquad$  **DAT405 Introduction to Data Science and AI, 2019-2020, Reading Period 3** <br />
$\qquad$ $\qquad$$\qquad$                   **Assignment 5: Reinforcement learning and Classification 2** <br />
$\qquad$ $\qquad$$\qquad$                   **Grader: Newton** <br />
$\qquad$ $\qquad$$\qquad$                     **Due Date: 23rd February, 23:59** <br />
$\qquad$ $\qquad$$\qquad$                   **Submitted by: Name, Personal No., Email** <br />


---


General guidelines:
*   All solutions to theoretical and pratical problems must be submitted in this ipynb notebook, and equations wherever required, should be formatted using LaTeX math-mode.
*   All discussion regarding practical problems, along with solutions and plots should be specified in this notebook. All plots/results should be visible such that the notebook do not have to be run. But the code in the notebook should reproduce the plots/results if we choose to do so.
*   Your name, personal number and email address should be specified above.
*   All tables and other additional information should be included in this notebook.
*   Before submitting, make sure that your code can run on another computer. That all plots can show on another computer including all your writing. It is good to check if your code can run here: https://colab.research.google.com.


Self-check 
1. Have you answered all questions to the best of your ability? 
2. Anything else you can easily check? (details, terminology, arguments, commenting for code etc.?) 

Grading will be based on a qualitative assessment of each assignment. It is important to:
*	Present clear arguments
*	Present the results in a pedagogical way
*	Show understanding of the topics (e.g, write a pseudocode) 
*	Give correct solutions
*	Make sure that the code is well commented 

**Again, as mentioned in general guidelines, all code should be written here. And this same ipython notebook file (RLAssignment.ipynb) should be submitted with answers and code written in it. NO SEPERATE FILE SHALL BE ACCEPTED.** 


# Primer

## Decision Making
The problem of **decision making under uncertainty** (commonly known as **reinforcement learning**) can be broken down into
two parts. First, how do we learn about the world? This involves both the
problem of modeling our initial uncertainty about the world, and that of drawing conclusions from evidence and our initial belief. Secondly, given what we
currently know about the world, how should we decide what to do, taking into
account future events and observations that may change our conclusions?
Typically, this will involve creating long-term plans covering possible future
eventualities. That is, when planning under uncertainty, we also need to take
into account what possible future knowledge could be generated when implementing our plans. Intuitively, executing plans which involve trying out new
things should give more information, but it is hard to tell whether this information will be beneficial. The choice between doing something which is already
known to produce good results and experiment with something new is known
as the **exploration-exploitation dilemma**.

## The exploration-exploitation trade-off

Consider the problem of selecting a restaurant to go to during a vacation.Lets say the
best restaurant you have found so far was **Les Epinards**. The food there is
usually to your taste and satisfactory. However, a well-known recommendations
website suggests that **King’s Arm** is really good! It is tempting to try it out. But
there is a risk involved. It may turn out to be much worse than **Les Epinards**,
in which case you will regret going there. On the other hand, it could also be
much better. What should you do?
It all depends on how much information you have about either restaurant,
and how many more days you’ll stay in town. If this is your last day, then it’s
probably a better idea to go to **Les Epinards**, unless you are expecting **King’s
Arm** to be significantly better. However, if you are going to stay there longer,
trying out **King’s Arm** is a good bet. If you are lucky, you will be getting much
better food for the remaining time, while otherwise you will have missed only
one good meal out of many, making the potential risk quite small.

## Overview
* To make things concrete, we will first focus on decision making under **no** uncertainity, i.e, given we have a world model, we can calculate the exact and optimal actions to take in it. We shall first introduce **Markov Decision Process (MDP)** as the world model. Then we give one algorithm (out of many) to solve it.


* Next, we will work through one type of reinforcement learning algorithm called Q-learning. Q-learning is an algorithm for making decisions under uncertainity, where uncertainity is over the possible world model (here MDP). It will find the optimal policy for the **unknown** MDP, assuming we do infinite exploration.

## Markov Decision Process

Markov Decision Process (MDP) provides a mathematical framework for modeling decision-making. It is a discrete time (distinct points in time) stochastic (randomly determined) process.

MDPs are made up of 4 parts:  
S: Finite set of states (Ex: s<sub>1</sub>, s<sub>2</sub> ... s<sub>N</sub>)  
A: Finite set of actions (Ex: North, South, East, West)  
P<sub>a</sub>(s,s'): Probability that action *a* in state *s* at time *t* will lead to state *s'* at time *t + 1*  
R<sub>a</sub>(s,s'): Immediate reward received after moving from state *s* to state *s'* by action *a*

An agent acts in an MDP at time *t*, by taking certain action *a* in state *s*, going to state *s'*, and getting a reward *r* from the world. It then repeats the process for certain no. of times, either finite or infinite.

We also include a $5^{th}$ part in the description of an MDP called Gamma $\gamma$.  
$\gamma$: The discount factor between 0 (inclusive) and 1 (exclusive). This determines how much credit you want to give to the future. If you think that the future reward is as important as the current reward you would set this to 0.99999. If you don't care about the future rewards you would set this to 0 and you only care about the current reward. For example, if your discount factor is 0.8 and after 5 steps you get a reward of 4 the present value of that reward is $0.8^4 * 5$ or ~2.

An MDP is a collection of states such that each state has a selection of actions associated with them. With each state-action pair comes a reward *r* (can be 0). Define a policy function: $\pi: s \rightarrow a$, which tells which action to take at each state.
  
We now use the famous dynamic programming equation, also known as Bellman Equation, to define optimality in an MDP. The following equation defines what we call the **value function** of state *s* following some fixed policy $\pi$:  

$$V^\pi(s) = \sum_{s'} P_{\pi(s)}(s,s') [R_{\pi(s)}(s,s') + \gamma V^\pi(s')]$$

We call $V^\pi$ as the value of policy $\pi$.  
  
Now, to find the **optimal** policy you will need to find the action that gives the highest reward.  

$$V^*(s) = max_a \sum_{s'} P_a(s,s') [R_a(s,s') + \gamma V^*(s')]$$

A real world example would be an inventory control system. Your states would be the amount of items you have in stock. Your actions would be the amount to order. The discrete time would be the days of the month. The reward would be the profit.  

A major drawback of MDPs is called the "Curse of Dimensionality". This states that the more states/actions you have the more computational difficult it is to solve.   


## Question 1 (2 points)

For the first question of the notebook, we give a quick example of an MDP. We would to see if you can put the definitions above into practice.

**Question a**: Given the following deterministic MDP (you select North, you move North), what is the optimal policy (path with the most points)?
  
*Notes*:  
  * The number in the box is the reward.  
  * Once you hit the end you are done. (Absorbing state)
  * S is the starting point.  
  * F is the ending point.  
  * Use N for North, E for East, S for South, and W for West. Not all actions are available at each state, for example, you can't choose N and W at starting state, as there exists no valid next states in those directions.  
  * Pass the directions as a single string. Ex: ESWN will make a circle.  
  


| | | |
|----------|----------|---------|
|S|1|1|
|1 |0|1|  
|-1|-1|0|  
|0 |0|F|

Question b,c will attempt to firm up your knowledge of the parts of an MDP. Just remember that for a state denoted by (x,y), state N/E/S/W to that are (x,y-1),(x+1,y),(x,y+1),(x-1,y) respectively. We take (0,0) as the starting state S.

**Question b:** What is the probability of going from state (1,0) to state (2,0) using action E ? ( i.e,  $P_E((1,0),(2,0))$ )

**Question c:** What is the reward for moving from state (1,0) to state (2,0) ? ( i.e, $R_E((1,0),(2,0))$ )

## Answers Question 1

### 1 A) 

Optimal policy : __SENESSS__ 

This gives $V^*(s)= 4$. Which is our path with the most points. We could start moving back and forth for an example state (0,1) and (0,2) before moving to F. This would however increase the total number of actions until the ending point. For that reason we decided to follow the following path: 

__Start S to End F:__
<br>
(0,0): S, V*(s) = 0
<br>
(0,0): action -> S, V*(s) = 0 + 1
<br>
(1,0): action -> E, V*(s) = 1 + 0
<br>
(1,1): action -> N, V*(s) = 1 + 1
<br>
(0,1): action -> E, V*(s) = 2 + 1
<br>
(0,2): action -> S, V*(S) = 3 + 1
<br>
(1,2): action -> S, V*(S) = 4 + 0
<br>
(2,2): action -> S
<br>
(3,2): F


As according to the optimal policy equation where we want to always choose the action which gives the highest reward. In our case $P_a=1$ since MDP is defined as deterministic.

$V^*(s) = max_a \sum_{s'} P_a(s,s') [R_a(s,s') + \gamma V^*(s')]$

#### We get $V^*(s)= 4$

***

### 1 B) 

The probability of going from state (1,0) to state (2,0) using action E is 100%. Action E is defined by having precondition (x,y) and postcondition (x+1,y), which can be translated into moving one step to the right in our MDP. If you use E when current state is (1,0) the current state will become (2,0).
<br> 

$\implies P_E((1,0),(2,0)) = 1 $

***

## 1 C)

The reward can be read from the MDP. When an action is made the reward given is whatever value is in the state it moves immediately after the action. And since the value 1 is in state (2,0) the answer is 1.

$\implies R_E((1,0),(2,0))=1$

***

## Value Iteration

The value iteration is one algorithm that can be used to find the optimal policy ($\pi^*$). Note that for any policy $\pi^*$ to be optimal, it must satisfy the Bellman equation for optimal value function $V^*$. For any candidate $V^*$, it must be such that plugging it in the RHS (right-hand-side) of Bellman equation should give the same $V^*$ again (by the recursive nature of this equation). This property will form the basis of our algorithm. Essentially, due to certain mathematical results, repeated application of RHS to any intial value function $V^0(s)$ will eventually lead to the value $V$ which statifies the Bellman equation. Hence repeated application of Bellman equation for optimal value function will also lead to optimal value function, we can then extract the optimal actions by simply noting the actions that satisfy the equation.    

The value function is based on the Bellman Equation for optimal value, which we recall here:  
$$V^*(s) = max_a \sum_{s'} P_a(s,s') [R_a(s,s') + \gamma V^*(s')]$$

Example: Below is a 3x3 grid. We are going to walk through a few iterations to firm up your understanding. Lets assume this time that success of taking any action is 0.8. Meaning if we take E from a valid state (x,y), we will go (x+1,y) 0.8 percent of time, but remain in same state the remaining time. We will have a discount factor ($\gamma$) of 0.9. Assume $V^0(s')=0$ for all s'. 

| | | |  
|----------|----------|---------|  
|0|0|0|
|0|10|0|  
|0|0|0|  


**Iteration 1**: It is trivial, V(s) becomes the $max_a \sum_{s'} P_a(s,s') R_a(s,s')$ since $V^0$ was zero for s'.

| | | |  
|----------|----------|---------|  
|0|8|0|
|8|2|8|  
|0|8|0|  
  
**Iteration 2**:  
  
Staring with cell (0,0): We find the expected value of each move:  
Action N: 0  
Action E: 0.8( 0 + 0.9 \* 8) + 0.2(0 + 0.9 \* 0) = 5.76  
Action S: 0.8( 0 + 0.9 \* 8) + 0.2(0 + 0.9 \* 0) = 5.76  
Action W: 0

Hence any action between E and S would have been best at this stage.

Similarly for cell (1,0):

Action S: 0.8( 10 + 0.9 \* 2) + 0.2(0 + 0.9 \* 8) = 10.88 (Action S is the maximizing action)  

Similar calculations for remaining cells give us:

| | | |  
|----------|----------|---------|  
|5.76|10.88|5.76|
|10.88|8.12|10.88|  
|5.76|10.88|5.76|  


## Question 2 (4 points)
Please code the value iteration algorithm just described here, and show the optimal value function of the above 3x3 grid problem at convergence.

### Answer to question 2 

In [8]:
from copy import deepcopy

#code implementation of the value iteration algoritm


#Create 3x3 grid with initilized values
vgrid = [[0,0,0],[0,0,0],[0,0,0]]
rgrid =[[0,0,0],[0,10,0],[0,0,0]]

#function to get every possible action from current state. Returns possible actions
def GetPossibleAction(row, col):
    actions = ["S","E","N","W"] #all actions
    if row == 0:
        actions.remove("N")
    if row == 2:
        actions.remove("S")
    if col ==  0:
        actions.remove("W")
    if col == 2:
        actions.remove("E")
    return actions


#function for one iteration of the values of the grid. Returns next grid
def Iterate(vgrid,rgrid, p, gamma):
    gridcopy = [[0 for i in range(len(vgrid)) ] for j in range(len(vgrid[0]))]
    for rows in range(len(vgrid)):
        for columns in range(len(vgrid[0])):
            points = [0,0,0,0]
            for a in GetPossibleAction(rows,columns):
                if a == "S":
                    points[0]=p*(rgrid[rows+1][columns] + gamma*vgrid[rows+1][columns])+(1-p)*(rgrid[rows][columns] + gamma*vgrid[rows][columns])
                if a == "E":
                    points[1]=p*(rgrid[rows][columns+1] + gamma*vgrid[rows][columns+1])+(1-p)*(rgrid[rows][columns] + gamma*vgrid[rows][columns])
                if a == "N":
                    points[2]=p*(rgrid[rows-1][columns] + gamma*vgrid[rows-1][columns])+(1-p)*(rgrid[rows][columns] + gamma*vgrid[rows][columns])
                if a == "W":
                    points[3]=p*(rgrid[rows][columns-1] + gamma*vgrid[rows][columns-1])+(1-p)*(rgrid[rows][columns] + gamma*vgrid[rows][columns])
            gridcopy[rows][columns]=round(max(points),2)
    vgrid = deepcopy(gridcopy)
    return vgrid


#function for iterating until the differene between state and next state becomes smaller than value epsilon. Returns optimal value function at convergence
def OptimalValue(vgrid,rgrid, eps):
    done = False
    while not done:
        prev_vgrid=deepcopy(vgrid)
        vgrid = deepcopy(Iterate(vgrid,rgrid,0.8,0.9)) 
        delta = []
        #check change of values for every state
        for rows in range(len(vgrid)):
            for columns in range(len(vgrid[0])):
                delta.append(abs(vgrid[rows][columns]-prev_vgrid[rows][columns]))
        if all(ele < eps for ele in delta): #we define that the value grid has converged when its change each iteration is less than eps
            done = True
    return vgrid

print(OptimalValue(vgrid,rgrid,0.0001))


[[45.6, 51.93, 45.6], [51.93, 48.03, 51.93], [45.6, 51.93, 45.6]]


### Question 2 reason

When running the above code we get: 

[[45.6, 51.93, 45.6], [51.93, 48.03, 51.93], [45.6, 51.93, 45.6]]

## Reinforcement Learning (RL)
Until now, we understood that knowing the MDP, specifically $P_a(s,s')$ and $R_a(s,s')$ allows us to efficiently find the optimal policy using value iteration algorithm, but RL or decision making under uncertainity arises from the question of making optimal decisions without knowing the true world model (MDP in this case).

So far we have defined the value of a state $V^\pi$, let us define the value of an action, $Q^\pi$:

$$Q^\pi(s,a) = \sum_{s'} P_a(s,s') [R_a(s,s') + \gamma V^\pi(s')]$$

i.e, the value of taking action *a* from state *s* and then following $\pi$ onwards. Similarly, the optimal Q-value equation is:

$$Q^*(s,a) = \sum_{s'} P_a(s,s') [R_a(s,s') + \gamma V^*(s')]$$

## Q-learning

Q-learning algorithm can be used by an agent unaware of its surroundings (unknown MDP). All it can do is take an action *a* at time *t* from state *s* and observe the reward *r* and next state *s'*, and repeat this process again. So how it can learn to act optimally under such uninformative conditions ? Answer is using Q-learning. Without going into its justification, we simply state the main-update rule of this algorithm below:

![alt text](https://chalmersuniversity.box.com/shared/static/5anbos4s9luoayb32jk6w3wy3w4jk3g3.png)

Where we simply maintain Q(s,a) value for each state-action pair in a table. It is proven to converge to the optimal policy of the underlying unknown MDP for certain values of learning rate $\alpha$. For our case, we set a constant $\alpha=0.1$.

## OpenAI Gym

We shall use already available simulators for different environments (world) using the popular OpenAI Gym library. It just implements [differnt types of simulators](https://gym.openai.com/) including ATARI games. Although here we will only focus on simple ones, such as [Chain enviroment](https://gym.openai.com/envs/NChain-v0/).
![alt text](https://chalmersuniversity.box.com/shared/static/6tthbzhpofq9gzlowhr3w8if0xvyxb2b.jpg)

## Question 3 (1 point)
Basically, there are 5 states, and two actions 'a' and 'b'. Each transition (s,a,s') is noted with its corresponding reward. You are to first familiarize with the framework using its [documentation](http://gym.openai.com/docs/), and then implement the Q-learning algorithm for the Chain enviroment (called 'NChain-v0') using default parameters. Finally print the $Q^*$ table at convergence. Take $\gamma=0.95$. You can refer to the Q-learning Jupyter notebook shown in class, uploaded on Canvas.




### Answer for question 3

In [16]:
import gym
import numpy as np
import random
import math
from copy import deepcopy

env = gym.make('NChain-v0')

#parameters in Q-learning algorithm
num_episodes = 10000
gamma = 0.95 #given
learning_rate = 0.1 #given
epsilon = 0.5#1 #0.15 #0.1

# initialize the Q table
Q = np.zeros([5, 2]) #5 states and 2 actions

#algorithm below copied from q_learning_frozen_lake.ipynb
#train q-table
for _ in range(num_episodes):
        state = env.reset()
        done = False
        while done == False:
            # First we select an action:
            if random.uniform(0, 1) < epsilon: # Flip a skewed coin
                action = env.action_space.sample() # Explore action space
            else:
                action = np.argmax(Q[state,:]) # Exploit learned values
            # Then we perform the action and receive the feedback from the environment
            new_state, reward, done, info = env.step(action)
            # Finally we learn from the experience by updating the Q-value of the selected action
            update = reward + (gamma*np.max(Q[new_state,:])) - Q[state, action]
            Q[state,action] += learning_rate*update 
            state = new_state
        #for printing q tables
        if _ > num_episodes-10 or _ < 10:
            print(_)
            print("------------------------")
            print(Q)





[2022-02-23 15:50:42,584] Making new env: NChain-v0
C:\Users\johan\anaconda3\lib\site-packages\gym\envs\registration.py:17: PkgResourcesDeprecationWarning: Parameters to load are deprecated.  Call .resolve and .require separately.
  result = entry_point.load(False)


0
------------------------
[[22.74890108 24.47289193]
 [17.84843112 23.56630062]
 [ 6.07627224 18.6909333 ]
 [ 9.81325114 16.66591996]
 [25.39490165 16.59533913]]
1
------------------------
[[27.44769102 28.36775213]
 [26.97582181 28.55015744]
 [20.87523204 28.67935894]
 [20.47161904 27.25630601]
 [42.69465197 21.85320601]]
2
------------------------
[[29.65683845 30.61222106]
 [29.86553497 31.17272693]
 [29.73289253 31.75619174]
 [37.75129136 31.55664805]
 [56.73380006 29.16521773]]
3
------------------------
[[44.20320977 41.9420208 ]
 [47.34494013 40.90974316]
 [49.39123022 43.09649468]
 [56.35631948 42.67133015]
 [66.63079766 42.82770638]]
4
------------------------
[[51.88440614 49.94256292]
 [55.44622253 51.54415454]
 [62.17335925 50.54648146]
 [70.73350526 52.17840027]
 [82.39898683 57.25265363]]
5
------------------------
[[59.89140762 58.89192476]
 [63.46789816 59.99908137]
 [65.91422855 60.69689578]
 [73.93734435 59.5589097 ]
 [76.85840865 64.94965224]]
6
--------------------

### Question 3 reasoning

When running the algorithm no convergance value seemed to be found. We tried comparing the difference between each value in every state to see if it ever fell below a certain constant. Running the algorithm with num_episodes = 10000 showed that the value never fell below a certain value. Instead we chose to display some episodes and found that the values seem to stabilize and give us an convergance interval. By only showing the last values in a 1000 or 10000 long episode we could see that the matrices were about the same. Giving us the conclusion that we have a convergance interval between ~50 to ~70. This can be seen in the txt file: RLAssignment-Q3-output.txt. Where we have printed the first 10 tables and the last 10 tables. In common for the last values is that the values for the first column (action "a") is greater than the value for the second column (action "b"). We intreped this as taking action "a" repeatedly will generate the biggest reward. The 10000th episode looks like: 

[[63.43240623 61.02015751]
<br>
 [67.23629168 61.71135264]
 <br>
 [70.77717075 65.58123666]
 <br>
 [74.9060682  64.23801194]
 <br>
 [88.35129618 64.15908349]]

***



## Question 4 (2 points)

a. Verify that the optimal $Q^*$ value obtained using Q-learning is same as the optimal value function $V^*$ for the corresponding MDP's optimal action. You would have to first define the MDP corresponding to Chain enviroment.

b. What is the importance of exploration in RL ? Explain with an example.

In [31]:
#find optimal value function V

from copy import deepcopy


vgrid = [0,0,0,0,0] #Should be 0 for all states before first iteration
rgrid =[2,0,0,0,10] #rewards for states in chain environment

actions = ["a","b"] #all actions are possible in every state



#function for one iteration of the values of the grid. Returns next grid
def Iterate(vgrid,rgrid, p, gamma):
    gridcopy = [0 for i in range(len(vgrid))]
    for state in range(len(vgrid)):
        points = [0]*len(actions)
        for a in actions: #propability of 0.2 that the agent "slips" and its action has the opposite effect
            if a == "a":
                if state == len(vgrid)-1:#if we are in the last state, "a" does not change current state
                    points[0]=p*(rgrid[state] + gamma*vgrid[state])+(1-p)*(rgrid[0] + gamma*vgrid[0])
                else:
                    points[0]=p*(rgrid[state+1] + gamma*vgrid[state+1])+(1-p)*(rgrid[0] + gamma*vgrid[0])
            if a == "b":
                try:
                    points[1]=p*(rgrid[0] + gamma*vgrid[0])+(1-p)*(rgrid[state+1] + gamma*vgrid[state+1])
                except IndexError:
                    points[1]=p*(rgrid[0] + gamma*vgrid[0])+(1-p)*(rgrid[state] + gamma*vgrid[state])
        gridcopy[state]=round(max(points),2)
    vgrid = deepcopy(gridcopy)
    return vgrid


#function for iterating until the differene between state and next state becomes smaller than value epsilon. Returns optimal value function at convergence
def optimalValue(vgrid,rgrid, eps):
    done = False
    while not done:
        prev_vgrid=deepcopy(vgrid)
        vgrid = deepcopy(Iterate(vgrid,rgrid,0.8,0.95)) #We assume gamma = 0.95 since it was 0.95 in Q3
        #print(vgrid)
        delta = []
        #check change of values for every state
        for state in range(len(vgrid)):
            delta.append(abs(vgrid[state]-prev_vgrid[state]))
        if all(ele < eps for ele in delta): #we define that the value grid has converged when its change each iteration is less than eps
            done = True
    return vgrid


print(optimalValue(vgrid,rgrid,0.0001))


[78.15, 82.77, 88.85, 96.85, 96.85]


In [44]:
#showing equality from the equation V*(S)=argmax_a Q*(s,a) 

#From the Q table obtained from Q3 from using Q-learning we can see that action "a" is the action that gives highest reward from every state.

#Therefore argmax_a Q*(s,a) is calculating the Q-value for doing action "a"

#--------------------------------------------------------------------------------------#

import gym
from copy import deepcopy

env = gym.make('NChain-v0')

#From assignment three we obtained the following Q table
Q = np.array([[63.43240623, 61.02015751],
    [67.23629168, 61.71135264],
    [70.77717075, 65.58123666],
    [74.9060682,  64.23801194],
    [88.35129618, 64.15908349]])

qgrid = [0,0,0,0,0] #Should be 0 for all states before first iteration
rgrid =[2,0,0,0,10] #rewards for states in chain environment

def Iterate(qgrid,rgrid, p, gamma):
    gridcopy = [0 for i in range(len(qgrid))]
    for state in range(len(qgrid)):
        points = 0 #to save value from the "best" action. Meaning the action that gives highest reward. 
        action = np.argmax(Q[state,:])
        if action == 0: #0 represents action "a"
            if state == len(qgrid)-1: #if we are in the last state, "a" does not change current state
                points=p*(rgrid[state] + gamma*qgrid[state])+(1-p)*(rgrid[0] + gamma*qgrid[0])
            else:
                points=p*(rgrid[state+1] + gamma*qgrid[state+1])+(1-p)*(rgrid[0] + gamma*qgrid[0])
        if action == 1: #1 represents action "b"
            try:
                points[1]=p*(rgrid[0] + gamma*qgrid[0])+(1-p)*(rgrid[state+1] + gamma*qgrid[state+1])
            except IndexError:
                points[1]=p*(rgrid[0] + gamma*qgrid[0])+(1-p)*(rgrid[state] + gamma*qgrid[state])
        gridcopy[state]=round(points,2)
    qgrid = deepcopy(gridcopy)
    return qgrid

    #function for iterating until the differene between state and next state becomes smaller than value epsilon. Returns optimal Q function at convergence
def optimalQ(qgrid,rgrid, eps):
    done = False
    while not done:
        prev_qgrid=deepcopy(qgrid)
        qgrid = deepcopy(Iterate(qgrid,rgrid,0.8,0.95)) #We assume gamma = 0.95 since it was 0.95 in Q3
        #print(vgrid)
        delta = []
        #check change of values for every state
        for state in range(len(qgrid)):
            delta.append(abs(qgrid[state]-prev_qgrid[state]))
        if all(ele < eps for ele in delta): #we define that the value grid has converged when its change each iteration is less than eps
            done = True
    return qgrid


print(optimalQ(qgrid,rgrid,0.0001))


    

[2022-02-23 18:51:44,722] Making new env: NChain-v0


[78.15, 82.77, 88.85, 96.85, 96.85]


### Question 4 reasoning 

#### 4 A) 

V*:[78.15, 82.77, 88.85, 96.85, 96.85]

Q*:[78.15, 82.77, 88.85, 96.85, 96.85]



#### 4 B)

The importance of exploration in reinforcement learning can be described as the importance to look further ahead. For an example if we were studying mice following a maze where in each turn the mouse can either go left or right. The mice will probably choose the path in which there is a piece of cheese laying (the reward). However what if that is the only piece of cheese there is going further along that path? If instead we could program mice to sometimes (say maybe 10% exploration) take the other path. Along this path there are several more rewards layin along the trail. If the mouse wasn't programmed to sometimes explore a little, the mouse would probably always take the path which immediatly gives an reward. After some time the mouse would (hopefully) learn to take the other path since it's more rewarding in the long run.  

## Question 5 (1 point)

Briefly discuss the k-armed bandit problem formulation and it's distinguishing feature as a special case of the reinforcement learning problem formulation.

### Answer Quesion 5 

The k-armed bandit problem comes initially from a gambler who tries do decide which slot machine to play on (armed bandits). The problem entails the trade off between exploration and exploitaition. In this problem the gambler has to choose which slot machine to play on. Each slot machine has a probability of giving a reward (getting 3 in a row). The gambler then has to decide wheter to keep playing on the machine his already playing on (exploitaition) or to use another machine (exploration) to be able to maximise his profits. This tradeoff has to be calculated since the gambler wants to maximise his profits, exploring a new machine could end in a loss but also gives new knowledge in regards to this new machine and the machines hes already played. By using his prior knowledge and adding in exploration the gambler wants to calculate the perfect tradeoff between exploitation and exploration; which machines to play and how many times he should pull the lever on the armed bandit.

## Note

* Until now, we have described algorithms for when no. of states and actions are finite. In coming weeks, you will be taught how to extend these methods to continous state enviroments like ATARI games.

# References
Primer/text based on the following references:
* http://www.cse.chalmers.se/~chrdimi/downloads/book.pdf
* https://github.com/olethrosdc/ml-society-science/blob/master/notes.pdf